### ISSUE TITLE
[Bug]: The node stops working after a transaction is sent

### ISSUE THREAD

**Issue by tomyrd** [0xPolygonMiden/miden-node#487]
### Version

next

### Other packages versions

miden-client: next (more specifically on commit `2f8707faeded3725a1d732310f616e3e2170a50d`)

### What happened?

The node works correctly until a transaction from the client is sent and then it "freezes", all of the requests after that don't get a response. No error message is printed in the node console.

We noticed that after reverting the commit `[chore: update dependencies (#476)]` the node didn't have this problem. More specifically if we only revert the commit [Update tonic et al](https://github.com/0xPolygonMiden/miden-node/pull/476/commits/636c6ab1065c0bacc2bc3945eec368527936989f) the node also was suddenly fixed. It might be a problem with the new versions of prost and tonic (the client in next is also using these new versions but it seems that it doesn't cause any issues).

### What should have happened?

The node should have processed the transaction and continue responding.

### How can this be reproduced?

With a node (from branch next) running locally. Run the command `make integration-test` in from the root of the `miden-client` repo (also in branch next).

### Relevant log output

```shell
The last log printed is always the same:


2024-09-06T16:35:45.727004Z  INFO store:get_block_inputs: miden-store: crates/store/src/server/api.rs:383: new
2024-09-06T16:35:45.727019Z  INFO store:get_block_inputs:validate_nullifiers: miden-store: crates/store/src/server/api.rs:591: new
2024-09-06T16:35:45.727028Z  INFO store:get_block_inputs:validate_nullifiers: miden-store: crates/store/src/server/api.rs:591: close, time.busy: 1.42µs, time.idle: 7.83µs
2024-09-06T16:35:45.727038Z  INFO store:get_block_inputs:validate_notes: miden-store: crates/store/src/server/api.rs:601: new
2024-09-06T16:35:45.727046Z  INFO store:get_block_inputs:validate_notes: miden-store: crates/store/src/server/api.rs:601: close, time.busy: 1.04µs, time.idle: 6.62µs
2024-09-06T16:35:45.727058Z  INFO store:get_block_inputs:select_block_header_by_block_num: miden-store: crates/store/src/db/mod.rs:249: new
2024-09-06T16:35:45.727281Z  INFO store:get_block_inputs:select_block_header_by_block_num: miden-store: crates/store/src/db/mod.rs:249: close, time.busy: 22.5µs, time.idle: 201µs
2024-09-06T16:35:45.727317Z  INFO store:get_block_inputs:select_note_inclusion_proofs: miden-store: crates/store/src/db/mod.rs:370: new
2024-09-06T16:35:45.727395Z  INFO store:get_block_inputs:select_note_inclusion_proofs: miden-store: crates/store/src/db/mod.rs:370: close, time.busy: 12.6µs, time.idle: 65.4µs
2024-09-06T16:35:45.727409Z  INFO store:get_block_inputs:select_block_headers: miden-store: crates/store/src/db/mod.rs:265: new
2024-09-06T16:35:45.727493Z  INFO store:get_block_inputs:select_block_headers: miden-store: crates/store/src/db/mod.rs:265: close, time.busy: 12.3µs, time.idle: 72.2µs
2024-09-06T16:35:45.727517Z  INFO store:get_block_inputs: miden-store: crates/store/src/server/api.rs:383: close, time.busy: 191µs, time.idle: 322µs


Then it just stops responding and no more logs are printed
```

**Comment by igamigo** [comment]
To add a little context after doing some tests:

- Running any transaction seems to stall the node indefinitely. Running the CLI for example and minting a note crashes the node as well.
- There seems to be some recursion problem. If you run the node in testing (@tomyrd was running it on release), you get a stack overflow:

```
2024-09-06T17:06:53.128378Z  INFO store:get_block_inputs: miden-store: crates/store/src/server/api.rs:383: close, time.busy: 1.76ms, time.idle: 3.39ms

thread 'tokio-runtime-worker' has overflowed its stack
fatal runtime error: stack overflow
/bin/sh: line 1: 51404 Abort trap: 6           cargo run --bin miden-node --features "testing" -- start --config ../tests/config/miden-node.toml node
make: *** [start-node] Error 134
```

**Comment by bobbinth** [comment]
oh - wow! interesting! Let's switch the client to `tomyrd-next-dependecies-without-dep-updates` for now and we can figure this out on Monday.

**Comment by igamigo** [comment]
I haven't tested whether this happens by using the faucet (although I think it should), but the faucet would also likely reproduce this issue. Maybe there should be an integration test that involves the faucet and node (or maybe a similar ad-hoc client).

**Comment by Mirko-von-Leipzig** [comment]
I've slowly narrowed this down some. Any non-empty block will likely encounter this freeze.

The block-producer never receives the `get_block_inputs` response from the store, which waits indefinitely at this stage halting all further batch/block production. I've reproduced this with only a single account ID in the input request. _Interestingly this does not occur for an empty request i.e. all empty vectors_.

The store api does correctly populate the data in the api trait - so it gets lost somewhere in the grpc machinery/network afaict.

At this stage its still unclear whether its due to the block-producer client or if the store never actually sends anything.

**Comment by Mirko-von-Leipzig** [comment]
Alright some progress of sorts. Called `get_note_auth_info` works as expected and I suspect most other store methods should work as expected.

The culprit appears to be the `account_states` in the response even though it only contains a single element for this test. If I clear this property (i.e. empty vec) then the call succeeds and life carries on as normal. Digging further into the inner properties.

**Comment by Mirko-von-Leipzig** [comment]
A single `proof` is enough to trigger this issue. The following does not work:

```rust
AccountBlockInputRecord {
  account_id: Some(0x288f490a6f861f5),
  account_hash: Some(0),
  proof: Some(MerklePath {
    siblings: [0]
  })
}
```

**Comment by Mirko-von-Leipzig** [comment]
This is getting weird. I think its consuming/dead-locking the entire tokio task pool.

If I spawn a task with a 4s sleep and a log just prior to making the block inputs request then that task never logs anything. `top` also shows the node as using 100% cpu.

I've dug into the `trace` level logs but tbh its just `hyper` related http frame logs, handshaking and other things which look normal but do repeat so something does appear stuck.

**Comment by Mirko-von-Leipzig** [comment]
Eugh. Alright solved and fixed in #490 - own goal by myself.

Some additional notes:

- The client integration test (`make integration-test`) always recompiles which was a bit frustrating until I had a separate test example going (takes ~20s)
- Our logs could use some work. In particular, its difficult to understand which rpc logs come from which component and which method. It would also be good to trace a single request between components, maybe by using `X-Request-ID` header and adding that to the trace. This would let us trace a client request -> rpc -> block-producer -> store in a sensible manner.
- We should probably be using timeouts
- Consider replacing `try_into`/`try_from` with named constructors instead. They're fairly opaque.
